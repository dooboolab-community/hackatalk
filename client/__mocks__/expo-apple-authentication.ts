export {
  AppleAuthenticationScope,
  AppleAuthenticationCredential,
} from 'expo-apple-authentication';

export const isAvailableAsync = jest.fn(() => Promise.resolve(true));
export const signInAsync = jest.fn(() =>
  Promise.resolve({
    user: 'uniqueUserId',
    identityToken: 'identityToken',
    realUserStatus: 1,
    authorizationCode: 'authorizationCode',
    fullName: {
      namePrefix: null,
      givenName: 'GivenName',
      nameSuffix: null,
      nickname: null,
      familyName: 'FamilyName',
      middleName: null,
    },
    email: 'email@privaterelay.appleid.com',
    state: null,
  }),
);
