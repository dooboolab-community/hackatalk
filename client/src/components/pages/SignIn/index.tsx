import * as AppleAuthentication from 'expo-apple-authentication';
import * as Crypto from 'expo-crypto';
import * as Device from 'expo-device';

import {
  Alert,
  Dimensions,
  Image,
  Platform,
  TouchableOpacity,
  View,
} from 'react-native';
import Animated, {
  block,
  clockRunning,
  cond,
  not,
  set,
  useCode,
} from 'react-native-reanimated';
import {AuthPayload, User} from '../../../types/graphql';
import {Button, EditText, ThemeType, useTheme} from 'dooboo-ui';
import {
  IC_LOGO_D,
  IC_LOGO_W,
  SvgApple,
  SvgFacebook,
  SvgGoogle,
} from '../../../utils/Icons';
import React, {FC, useEffect, useState} from 'react';
import type {
  UserSignInAppleMutation,
  UserSignInAppleMutationResponse,
} from '../../../__generated__/UserSignInAppleMutation.graphql';
import type {
  UserSignInEmailMutation,
  UserSignInEmailMutationResponse,
} from '../../../__generated__/UserSignInEmailMutation.graphql';
import {delay, spring, useClock, useValue} from 'react-native-redash';
import {showAlertForError, validateEmail} from '../../../utils/common';
import {signInEmail, signInWithApple} from '../../../relay/queries/User';
import styled, {css} from '@emotion/native';

import AsyncStorage from '@react-native-async-storage/async-storage';
import {AuthStackNavigationProps} from '../../navigations/AuthStackNavigator';
import type {NotificationCreateNotificationMutation} from '../../../__generated__/NotificationCreateNotificationMutation.graphql';
import SocialSignInButton from './SocialSignInButton';
import StatusBar from '../../uis/StatusBar';
import {createNotification} from '../../../relay/queries/Notification';
import {getString} from '../../../../STRINGS';
import {useAuthContext} from '../../../providers/AuthProvider';
import {useMutation} from 'react-relay';
import {useNavigation} from '@react-navigation/core';

const AnimatedTouchableOpacity =
  Animated.createAnimatedComponent(TouchableOpacity);

const Container = styled.SafeAreaView`
  flex: 1;
  justify-content: center;
  background: ${({theme}) => theme.background};
`;

const Wrapper = styled.View`
  margin: 40px;
`;

const LogoWrapper = styled.View`
  margin-top: 44px;
  margin-bottom: 72px;
`;

const StyledLogoText = styled.Text`
  align-self: flex-start;
  color: ${({theme}) => theme.text};
  font-size: 20px;
  font-weight: bold;
  margin-left: 6px;
`;

const ButtonWrapper = styled.View`
  margin-top: 12px;
  height: 52px;
  width: 100%;
  flex-direction: row;
`;

const FindPwTouchOpacity = styled.TouchableOpacity`
  padding: 20px;
  margin-bottom: 44px;
  align-self: center;
`;

const FindPwText = styled.Text`
  color: ${({theme}) => theme.link};
  text-decoration-line: underline;
`;

const SocialButtonWrapper = styled.View`
  margin-bottom: 24px;
`;

const StyledAgreementTextWrapper = styled.View`
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  padding: 0 0 40px 0;
`;

const StyledAgreementText = styled.Text`
  line-height: 22px;
  color: #777;
`;

const StyledAgreementLinedText = styled.Text`
  line-height: 22px;
  color: ${({theme}) => theme.link};
  text-decoration-line: underline;
`;

const StyledScrollView = styled.ScrollView`
  align-self: center;
  width: 100%;

  /* ${({theme: {isDesktop}}) =>
    isDesktop &&
    css`
      width: 50%;
      max-width: 800;
    `} */
`;

const SignIn: FC = () => {
  const navigation = useNavigation<AuthStackNavigationProps<'SignIn'>>();
  const {setUser} = useAuthContext();
  const {theme, changeThemeType, themeType} = useTheme();

  const [signingInApple, setSigningInApple] = useState<boolean>(false);
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [errorEmail, setErrorEmail] = useState<string>('');
  const [errorPassword, setErrorPassword] = useState<string>('');

  const [commitEmail, isInFlight] =
    useMutation<UserSignInEmailMutation>(signInEmail);

  const [commitApple, isAppleInFlight] =
    useMutation<UserSignInAppleMutation>(signInWithApple);

  const [commitNotification] =
    useMutation<NotificationCreateNotificationMutation>(createNotification);

  const createNotificationIfPushTokenExists = async (): Promise<void> => {
    const pushToken = await AsyncStorage.getItem('push_token');

    if (pushToken) {
      const createNotificationMutationConfig = {
        variables: {
          token: pushToken,
          device: Device.modelName,
          os: Device.osName,
        },
      };

      commitNotification(createNotificationMutationConfig);
    }
  };

  const goToSignUp = (): void => {
    navigation.navigate('SignUp');
  };

  const goToFindPw = (): void => {
    navigation.navigate('FindPw');
  };

  const signIn = async (): Promise<void> => {
    if (!validateEmail(email)) {
      setErrorEmail(getString('EMAIL_FORMAT_NOT_VALID'));

      return;
    }

    if (!password) {
      setErrorPassword(getString('PASSWORD_REQUIRED'));

      return;
    }

    const licenseAgreed = JSON.parse(
      (await AsyncStorage.getItem('license_agreed')) as string,
    );

    if (!licenseAgreed) return navigation.navigate('LicenseAgreement');

    const mutationConfig = {
      variables: {
        email,
        password,
      },

      onCompleted: (response: UserSignInEmailMutationResponse) => {
        const {token, user} = response.signInEmail as AuthPayload;

        if (user && !user.verified)
          return navigation.navigate('VerifyEmail', {
            email,
          });

        AsyncStorage.setItem('token', token);

        createNotificationIfPushTokenExists();

        setUser(user as User);
      },

      onError: (error: Error): void => {
        setErrorPassword(error.message);
        setErrorPassword(getString('PASSWORD_INCORRECT'));
      },
    };

    commitEmail(mutationConfig);
  };

  const goToWebView = (uri: string): void => {
    navigation.navigate('WebView', {uri});
  };

  const appleLogin = async (): Promise<void> => {
    setSigningInApple(true);

    try {
      const csrf = Math.random().toString(36).substring(2, 15);
      const nonce = Math.random().toString(36).substring(2, 10);

      const hashedNonce = await Crypto.digestStringAsync(
        Crypto.CryptoDigestAlgorithm.SHA256,
        nonce,
      );

      const appleCredential = await AppleAuthentication.signInAsync({
        requestedScopes: [
          AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
          AppleAuthentication.AppleAuthenticationScope.EMAIL,
        ],
        state: csrf,
        nonce: hashedNonce,
      });

      const {identityToken} = appleCredential;

      if (identityToken) {
        const mutationConfig = {
          variables: {
            accessToken: identityToken,
          },
          onCompleted: (response: UserSignInAppleMutationResponse) => {
            const {token, user} = response.signInWithApple as AuthPayload;

            AsyncStorage.setItem('token', token);

            createNotificationIfPushTokenExists();

            setUser(user as User);
          },
          onError: (err: Error) => {
            showAlertForError(err);
          },
        };

        commitApple(mutationConfig);
      }
    } catch (e) {
      if (e.code === 'ERR_CANCELED') {
        // handle that the user canceled the sign-in flow
      } else
        Platform.select({
          // @ts-ignore
          // eslint-disable-next-line no-alert
          web: alert(`Apple Login Error: ${e.code} - ${e.message}`),
          default: Alert.alert(`Apple Login Error: ${e.code} - ${e.message}`),
        });
    } finally {
      setSigningInApple(false);
    }
  };

  useEffect(() => {
    // console.log('appOwnership', Constants.appOwnership);
  }, []);

  const logoSize = 80;
  const logoTransformAnimValue = useValue(0);
  const {width: screenWidth, height: screenHeight} = Dimensions.get('window');

  const logoInitialPosition = {
    x: (screenWidth - logoSize) * 0.5,
    y: screenHeight * 0.5 - logoSize,
  };

  const logoFinalPosition = {
    x: 30,
    y: 80,
  };

  const logoScale = logoTransformAnimValue.interpolate({
    inputRange: [0, 1],
    outputRange: [2, 1],
  });

  const logoPositionX = logoTransformAnimValue.interpolate({
    inputRange: [0, 1],
    outputRange: [logoInitialPosition.x, logoFinalPosition.x],
  });

  const logoPositionY = logoTransformAnimValue.interpolate({
    inputRange: [0, 1],
    outputRange: [logoInitialPosition.y, logoFinalPosition.y],
  });

  const animating = useValue<number>(0);
  const clock = useClock();

  useCode(() => block([delay(set(animating, 1), 100)]), []);

  useCode(
    () =>
      block([
        cond(animating, [
          set(
            logoTransformAnimValue,
            spring({
              clock,
              to: 1,
              from: 0,
              config: {mass: 1.5, stiffness: 36},
            }),
          ),
        ]),
        cond(not(clockRunning(clock)), set(animating, 0)),
      ]),
    [],
  );

  return (
    <Container>
      <StatusBar />

      <StyledScrollView>
        <AnimatedTouchableOpacity
          testID="theme-test"
          onPress={(): void => changeThemeType()}
          style={{
            zIndex: 15,
            position: 'absolute',
            left: logoPositionX,
            top: logoPositionY,
            transform: [{scale: logoScale}],
          }}>
          <Image
            style={{width: logoSize, height: logoSize, resizeMode: 'cover'}}
            source={themeType === ThemeType.DARK ? IC_LOGO_D : IC_LOGO_W}
          />
        </AnimatedTouchableOpacity>
        <Wrapper>
          <LogoWrapper>
            <View style={{height: 12 + 60}} />
            <StyledLogoText>{getString('HELLO')}</StyledLogoText>
          </LogoWrapper>
          <EditText
            type="row"
            testID="input-email"
            styles={{
              container: {
                borderColor: theme.text,
              },
              input: {
                color: theme.text,
                height: 52,
              },
            }}
            style={{marginBottom: 20}}
            labelText={getString('EMAIL')}
            focusColor={theme.focused}
            placeholderTextColor={theme.placeholder}
            placeholder="hello@example.com"
            value={email}
            onChangeText={(text: string): void => {
              setEmail(text);
              setErrorEmail('');
            }}
            errorText={errorEmail}
            onSubmitEditing={signIn}
          />
          <EditText
            testID="input-password"
            type="row"
            styles={{
              container: {
                borderColor: theme.text,
              },
              input: {
                color: theme.text,
                height: 52,
              },
            }}
            style={{marginBottom: 20}}
            labelText={getString('PASSWORD')}
            focusColor={theme.focused}
            placeholder="******"
            placeholderTextColor={theme.placeholder}
            value={password}
            onChangeText={(text: string): void => {
              setPassword(text);
              setErrorPassword('');
            }}
            errorText={errorPassword}
            onSubmitEditing={signIn}
            secureTextEntry={true}
          />
          <ButtonWrapper>
            <Button
              testID="btn-sign-up"
              onPress={goToSignUp}
              style={{flex: 1}}
              styles={{
                container: {
                  backgroundColor: theme.btnPrimaryLight,
                  borderColor: theme.btnPrimary,
                  borderWidth: 1,

                  flex: 1,
                  flexDirection: 'row',
                  justifyContent: 'center',
                },
                text: {
                  color: theme.btnPrimary,
                  fontSize: 14,
                  fontWeight: 'bold',
                },
                hovered: {
                  borderColor: theme.text,
                },
              }}
              text={getString('SIGN_UP')}
            />
            <View style={{width: 12}} />
            <Button
              testID="btn-sign-in"
              loading={isInFlight}
              onPress={signIn}
              style={{
                flex: 1,
              }}
              styles={{
                container: {
                  height: 52,
                  backgroundColor: theme.primary,
                  flex: 1,
                  alignSelf: 'stretch',
                  borderWidth: 1,

                  flexDirection: 'row',
                  justifyContent: 'center',
                },
                text: {
                  color: theme.btnPrimaryFont,
                  fontSize: 14,
                  fontWeight: 'bold',
                },
                hovered: {
                  borderColor: theme.text,
                },
              }}
              text={getString('LOGIN')}
            />
          </ButtonWrapper>
          <FindPwTouchOpacity testID="btn-find-pw" onPress={goToFindPw}>
            <FindPwText>{getString('FORGOT_PW')}</FindPwText>
          </FindPwTouchOpacity>
          <SocialButtonWrapper>
            {Platform.select({
              ios: (
                <Button
                  testID="btn-apple"
                  style={{marginBottom: 12}}
                  styles={{
                    container: {
                      backgroundColor: theme.appleBackground,
                      borderColor: theme.appleText,
                      width: '100%',
                      height: 48,
                      borderWidth: 1,
                      borderRadius: 100,
                    },
                    text: {
                      fontWeight: '700',
                      color: theme.appleText,
                    },
                  }}
                  leftElement={
                    <View style={{marginRight: 6}}>
                      <SvgApple width={18} height={18} fill={theme.appleIcon} />
                    </View>
                  }
                  loading={signingInApple || isAppleInFlight}
                  indicatorColor={theme.primary}
                  onPress={appleLogin}
                  text={getString('SIGN_IN_WITH_APPLE')}
                />
              ),
            })}
            <SocialSignInButton
              svgIcon={
                <SvgFacebook width={18} height={18} fill={theme.facebookIcon} />
              }
              onUserCreated={(user?: User): void => {
                createNotificationIfPushTokenExists();
                setUser?.(user);
              }}
              socialProvider={'facebook'}
            />
            <SocialSignInButton
              svgIcon={
                <SvgGoogle width={20} height={20} fill={theme.googleIcon} />
              }
              onUserCreated={(user?: User): void => {
                createNotificationIfPushTokenExists();
                setUser?.(user);
              }}
              socialProvider="google"
            />
          </SocialButtonWrapper>
          <StyledAgreementTextWrapper>
            <StyledAgreementText>{getString('AGREEMENT1')}</StyledAgreementText>
            <StyledAgreementLinedText
              testID="btn-terms"
              onPress={(): void =>
                goToWebView('https://legacy.dooboolab.com/termsofservice')
              }>
              {getString('AGREEMENT2')}
            </StyledAgreementLinedText>
            <StyledAgreementText>{getString('AGREEMENT3')}</StyledAgreementText>
            <StyledAgreementLinedText
              testID="btn-privacy"
              onPress={(): void =>
                goToWebView('https://legacy.dooboolab.com/privacyandpolicy')
              }>
              {getString('AGREEMENT4')}
            </StyledAgreementLinedText>
            <StyledAgreementText>{getString('AGREEMENT5')}</StyledAgreementText>
          </StyledAgreementTextWrapper>
        </Wrapper>
      </StyledScrollView>
    </Container>
  );
};

export default SignIn;
