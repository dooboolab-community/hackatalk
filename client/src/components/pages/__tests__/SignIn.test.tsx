import * as AppleAuthentication from 'expo-apple-authentication';

import {MockPayloadGenerator, createMockEnvironment} from 'relay-test-utils';
import {act, fireEvent, render, waitFor} from '@testing-library/react-native';
import {
  createMockNavigation,
  createTestElement,
} from '../../../../test/testUtils';

import AsyncStorage from '@react-native-async-storage/async-storage';
import AuthContext from '../../../providers/AuthProvider';
import {FetchMock} from 'jest-fetch-mock';
import React from 'react';
import ReactNavigation from '@react-navigation/core';
import SignIn from '../SignIn';
import {ThemeType} from 'dooboo-ui';

const fetchMock = fetch as FetchMock;

fetchMock.mockResponse(JSON.stringify({id: 1}));

const mockNavigation = createMockNavigation();

jest.mock('@react-navigation/core', () => ({
  ...jest.requireActual<typeof ReactNavigation>('@react-navigation/core'),
  useNavigation: () => mockNavigation,
}));

jest.mock('../../../components/pages/SignIn/SocialSignInButton', () => 'test');

describe('[SignIn] rendering test', () => {
  it('should render without crashing', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);

    const json = screen.toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should render [Dark] mode without crashing', () => {
    const component = createTestElement(<SignIn />, {
      themeType: ThemeType.DARK,
    });

    const screen = render(component);

    const json = screen.toJSON();

    expect(json).toBeTruthy();
    expect(json).toMatchSnapshot();
  });
});

describe('[SignIn] interaction', () => {
  it('should change theme when icon is pressed', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);
    const themeTouch = screen.getByTestId('theme-test');

    fireEvent.press(themeTouch);
  });

  it('should invoke changeText event handler when email changed ', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);
    const textInput = screen.getByTestId('input-email');

    fireEvent.changeText(textInput, 'email@email.com');

    expect(textInput.props.value).toEqual('email@email.com');
  });

  it('should invoke changeText event handler when password changed ', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);

    const textInput = screen.getByTestId('input-password');

    fireEvent.changeText(textInput, 'pw test');

    expect(textInput.props.value).toEqual('pw test');
  });

  it('should simulate signUp button has clicked', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);
    const btnSignUp = screen.getByTestId('btn-sign-up');

    fireEvent.press(btnSignUp);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('SignUp');
  });

  it('should navigate to [FindPw] when button has pressed', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);
    const findPwBtn = screen.getByTestId('btn-find-pw');

    fireEvent.press(findPwBtn);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('FindPw');
  });

  it('should navigate to [WebView] when terms has been pressed', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);
    const btnTerms = screen.getByTestId('btn-terms');

    fireEvent.press(btnTerms);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('WebView', {
      uri: 'https://legacy.dooboolab.com/termsofservice',
    });
  });

  it('should navigate to [WebView] when terms has been pressed again', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);
    const btnPrivary = screen.getByTestId('btn-privacy');

    fireEvent.press(btnPrivary);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('WebView', {
      uri: 'https://legacy.dooboolab.com/privacyandpolicy',
    });
  });

  describe('onSignIn', () => {
    it('should call signIn when button has clicked and navigation switches to [MainStack]', async () => {
      const mockEnvironment = createMockEnvironment();

      const component = createTestElement(<SignIn />, {
        environment: mockEnvironment,
      });

      const screen = render(component);

      jest.spyOn(AsyncStorage, 'setItem').mockImplementation(jest.fn());

      jest
        .spyOn(AsyncStorage, 'getItem')
        .mockImplementation(jest.fn().mockResolvedValue(JSON.stringify(true)));

      jest.spyOn(AuthContext, 'useAuthContext').mockImplementation(() => ({
        state: {
          user: undefined,
        },
        setUser: jest.fn().mockReturnValue({
          id: 'userId',
          email: 'email@email.com',
          nickname: 'nickname',
          statusMessage: 'status',
        }),
      }));

      const textInput = screen.getByTestId('input-email');

      fireEvent.changeText(textInput, 'test@email.com');

      const passwordInput = screen.getByTestId('input-password');

      fireEvent.changeText(passwordInput, 'password');

      const btnSignIn = screen.getByTestId('btn-sign-in');

      fireEvent.press(btnSignIn);

      const operation = await waitFor(() =>
        mockEnvironment.mock.getMostRecentOperation(),
      );

      expect(operation).toBeTruthy();
    });

    it('should call signIn with invalid params and  check whether it catches error', async () => {
      const mockEnvironment = createMockEnvironment();

      const component = createTestElement(<SignIn />, {
        environment: mockEnvironment,
      });

      const screen = render(component);

      const textInput = screen.getByTestId('input-email');

      fireEvent.changeText(textInput, 'invalid@email.com');

      const passwordInput = screen.getByTestId('input-password');

      fireEvent.changeText(passwordInput, 'password');

      const btnSignIn = screen.getByTestId('btn-sign-in');

      fireEvent.press(btnSignIn);

      const operation = await waitFor(() =>
        mockEnvironment.mock.getMostRecentOperation(),
      );

      act(() => {
        mockEnvironment.mock.reject(operation, new Error('Uh-oh'));
      });

      // TODO: Test what happens after rejected operation.
    });
  });

  it('should call Apple signin when pressing button', async () => {
    const mockEnvironment = createMockEnvironment();
    const generatorSpy = jest.spyOn(MockPayloadGenerator, 'generate');

    const component = createTestElement(<SignIn />, {
      environment: mockEnvironment,
    });

    const screen = render(component);
    const btnApple = screen.getByTestId('btn-apple');

    fireEvent.press(btnApple);

    await waitFor(() => {
      mockEnvironment.mock.resolveMostRecentOperation((operation) =>
        MockPayloadGenerator.generate(operation),
      );
    });

    expect(generatorSpy).toHaveBeenCalledTimes(1);
  });
});

describe('Apple SignIn', () => {
  it('should Apple auth available', async () => {
    return expect(AppleAuthentication.isAvailableAsync()).resolves.toBeTruthy();
  });

  it('should signin with Apple when [btn-apple] has pressed', async () => {
    const component = createTestElement(<SignIn />);
    const screen = render(component);
    const btnApple = screen.getByTestId('btn-apple');

    fireEvent.press(btnApple);

    const signInResult = await waitFor(() =>
      AppleAuthentication.signInAsync({
        requestedScopes: [
          AppleAuthentication.AppleAuthenticationScope.EMAIL,
          AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
        ],
      }),
    );

    expect(signInResult).toMatchObject({
      user: 'uniqueUserId',
      identityToken: 'identityToken',
      realUserStatus: 1,
      authorizationCode: 'authorizationCode',
      fullName: {
        namePrefix: null,
        givenName: 'GivenName',
        nameSuffix: null,
        nickname: null,
        familyName: 'FamilyName',
        middleName: null,
      },
      email: 'email@privaterelay.appleid.com',
      state: null,
    });
  });
});
