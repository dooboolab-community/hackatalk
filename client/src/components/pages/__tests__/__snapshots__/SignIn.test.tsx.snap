// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SignIn] rendering test should render [Dark] mode without crashing 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#000",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <RCTScrollView
      style={
        Array [
          Object {
            "alignSelf": "center",
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": undefined,
              "left": undefined,
              "opacity": 1,
              "position": "absolute",
              "top": undefined,
              "width": undefined,
              "zIndex": 15,
            }
          }
          testID="theme-test"
        >
          <Image
            source={
              Object {
                "canInstrument": true,
                "createTransformer": [Function],
                "getCacheKey": [Function],
                "process": [Function],
              }
            }
            style={
              Object {
                "height": "100%",
                "resizeMode": "cover",
                "width": "100%",
              }
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "marginBottom": 40,
                "marginLeft": 40,
                "marginRight": 40,
                "marginTop": 40,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "marginBottom": 72,
                  "marginTop": 44,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Object {
                  "height": 72,
                }
              }
            />
            <Text
              style={
                Array [
                  Object {
                    "alignSelf": "flex-start",
                    "color": "white",
                    "fontSize": 20,
                    "fontWeight": "bold",
                    "marginLeft": 6,
                  },
                  undefined,
                ]
              }
            >
              Hello
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                  "flexDirection": "column",
                },
                Object {
                  "marginBottom": 4,
                },
              ]
            }
            testID="container-id"
          >
            <View
              style={
                Array [
                  Array [
                    Object {
                      "alignItems": "center",
                      "alignSelf": "stretch",
                      "borderBottomColor": "white",
                      "borderBottomWidth": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    },
                    Object {
                      "borderColor": "white",
                    },
                  ],
                  false,
                  Object {
                    "borderColor": "white",
                  },
                  Object {
                    "borderBottomColor": "white",
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Array [
                      Object {
                        "color": "#6D6D6D",
                        "fontSize": 14,
                        "marginRight": 2,
                      },
                      Object {
                        "width": 80,
                      },
                    ],
                    false,
                    Object {},
                  ]
                }
              >
                Email
              </Text>
              <TextInput
                autoCapitalize="none"
                editable={true}
                multiline={false}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                onSubmitEditing={[Function]}
                placeholder="hello@example.com"
                placeholderTextColor="#6D6D6D"
                secureTextEntry={false}
                style={
                  Array [
                    Array [
                      Object {
                        "color": "white",
                        "flex": 1,
                        "fontSize": 14,
                        "paddingHorizontal": 8,
                        "paddingVertical": 10,
                      },
                      Object {
                        "color": "white",
                        "height": 52,
                      },
                    ],
                    false,
                  ]
                }
                testID="input-email"
                value=""
              />
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  Array [
                    Array [
                      Object {
                        "color": "#FF002E",
                        "fontSize": 12,
                        "marginTop": 4,
                      },
                      undefined,
                    ],
                    Object {
                      "color": "#FF002E",
                      "flex": 1,
                    },
                  ]
                }
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                  "flexDirection": "column",
                },
                Object {
                  "marginBottom": 20,
                },
              ]
            }
            testID="container-id"
          >
            <View
              style={
                Array [
                  Array [
                    Object {
                      "alignItems": "center",
                      "alignSelf": "stretch",
                      "borderBottomColor": "white",
                      "borderBottomWidth": 1,
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    },
                    Object {
                      "borderColor": "white",
                    },
                  ],
                  false,
                  Object {
                    "borderColor": "white",
                  },
                  Object {
                    "borderBottomColor": "white",
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Array [
                      Object {
                        "color": "#6D6D6D",
                        "fontSize": 14,
                        "marginRight": 2,
                      },
                      Object {
                        "width": 80,
                      },
                    ],
                    false,
                    Object {},
                  ]
                }
              >
                Password
              </Text>
              <TextInput
                autoCapitalize="none"
                editable={true}
                multiline={false}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                onSubmitEditing={[Function]}
                placeholder="******"
                placeholderTextColor="#6D6D6D"
                secureTextEntry={true}
                style={
                  Array [
                    Array [
                      Object {
                        "color": "white",
                        "flex": 1,
                        "fontSize": 14,
                        "paddingHorizontal": 8,
                        "paddingVertical": 10,
                      },
                      Object {
                        "color": "white",
                        "height": 52,
                      },
                    ],
                    false,
                  ]
                }
                testID="input-password"
                value=""
              />
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <Text
                style={
                  Array [
                    Array [
                      Object {
                        "color": "#FF002E",
                        "fontSize": 12,
                        "marginTop": 4,
                      },
                      undefined,
                    ],
                    Object {
                      "color": "#FF002E",
                      "flex": 1,
                    },
                  ]
                }
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "height": 52,
                  "marginTop": 4,
                  "width": "100%",
                },
                undefined,
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                  "opacity": 1,
                }
              }
              testID="btn-sign-up"
            />
            <View
              style={
                Object {
                  "width": 12,
                }
              }
            />
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                  "opacity": 1,
                }
              }
              testID="btn-sign-in"
            />
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignSelf": "center",
                "marginBottom": 28,
                "opacity": 1,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
              }
            }
            testID="btn-find-pw"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#8BC9FF",
                    "textDecorationLine": "underline",
                  },
                  undefined,
                ]
              }
            >
              Forgot password?
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginBottom": 24,
                },
                undefined,
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "marginBottom": 12,
                  "opacity": 1,
                }
              }
              testID="btn-apple"
            />
            <test
              onUserCreated={[Function]}
              socialProvider="facebook"
              svgIcon={
                <SvgMock
                  fill="#345997"
                  height={18}
                  width={18}
                />
              }
            />
            <test
              onUserCreated={[Function]}
              socialProvider="google"
              svgIcon={
                <SvgMock
                  fill="#E04238"
                  height={20}
                  width={20}
                />
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "center",
                  "paddingBottom": 40,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 0,
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 14,
                  },
                  Object {
                    "textAlign": "center",
                  },
                ]
              }
            >
              We consider that you agree with 
              <Text
                onPress={[Function]}
                style={
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 14,
                    },
                    Object {
                      "textDecorationLine": "underline",
                    },
                  ]
                }
              >
                Terms for agreement
              </Text>
               and 
              <Text
                onPress={[Function]}
                style={
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 14,
                    },
                    Object {
                      "textDecorationLine": "underline",
                    },
                  ]
                }
              >
                Privacy and Policy
              </Text>
               by going onto next step.
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</RNCSafeAreaProvider>
`;
